hyper_function.h

```diff
@@ -1,11 +1,25 @@
 #pragma once
 
+#if USE_STDFUNCTION
```

Comment:
I'd be interested to see what error message you're getting from the compiler when `ISSTEAM && !USE_STDFUNCTION`. Maybe it's something easily fixable. Even though I see a huge speedup from `hyper_function`, I don't think it makes sense to keep it if it requires maintaining two different codepaths in perpetuity.
It would also be really nice to get the `ISSTEAM` build running on TravisCI (e.g. the way Emscripten is, using a Docker image), but I guess that might not be possible?

Orthogonally, you're right, the difference in compile time between `std::function` and `hyper_function` seems to be Clang-specific. E.g. I see a big difference between [202](https://travis-ci.org/zenorogue/hyperrogue/jobs/483899113) and [151](https://travis-ci.org/zenorogue/hyperrogue/jobs/490646163) seconds (Clang/Linux) but no difference (or worse than no difference) between [157](https://travis-ci.org/zenorogue/hyperrogue/jobs/483899114) and [173](https://travis-ci.org/zenorogue/hyperrogue/jobs/490646164) seconds (GCC/Linux).

Reply:
Yes, probably they are easy to fix. I get the following errors:
```
hyper_function.h:13:44: error: 'virtual hyper_function_state_base<R, Args>::~hyper_function_state_base()' declared virtual cannot be defaulted in the class body
hyper_function.h:18:11: error: expected nested-name-specifier before 'Self'
hyper_function.h:18:11: error: using-declaration for non-member at class scope
hyper_function.h:18:16: error: expected ';' before '=' token
hyper_function.h:18:16: error: expected unqualified-id before '=' token
hyper_function.h:21:26: error: expected ';' at end of member declaration
hyper_function.h:21:32: error: 'override' does not name a type
hyper_function.h:24:52: error: expected ';' at end of member declaration
hyper_function.h:24:58: error: 'override' does not name a type
hyper_function.h:32:51: sorry, unimplemented: non-static data member initializers
hyper.h:22:21: error: expected unqualified-id before 'using'
```

Reply:
Yes, probably they are easy to fix. I get the following errors:
```
hyper_function.h:13:44: error: 'virtual hyper_function_state_base<R, Args>::~hyper_function_state_base()' declared virtual cannot be defaulted in the class body
hyper_function.h:18:11: error: expected nested-name-specifier before 'Self'
hyper_function.h:18:11: error: using-declaration for non-member at class scope
hyper_function.h:18:16: error: expected ';' before '=' token
hyper_function.h:18:16: error: expected unqualified-id before '=' token
hyper_function.h:21:26: error: expected ';' at end of member declaration
hyper_function.h:21:32: error: 'override' does not name a type
hyper_function.h:24:52: error: expected ';' at end of member declaration
hyper_function.h:24:58: error: 'override' does not name a type
hyper_function.h:32:51: sorry, unimplemented: non-static data member initializers
hyper.h:22:21: error: expected unqualified-id before 'using'
```

Reply:
Looks like GCC 4.6.
So the Steam build must not use either `Makefile.simple` or `Makefile.am` right now, because those makefiles both expect `-std=c++11` to be supported, whereas GCC 4.6 requires `-std=c++0x`, is that right?
I've got a branch at `Quuxplusone/steam` [(TravisCI)](https://travis-ci.org/Quuxplusone/hyperrogue/builds/491355022) which addresses the issues with GCC 4.6.4 and starts testing GCC 4.6.4 on Travis. Is that useful to you? and is there anything I could add to that build (e.g. `-DISSTEAM=1`) that would make it even more "realistic"?